projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/intrinsics/interop/LLVMTruffleImportCached.java

```diff
@@ -0,0 +1,103 @@
+ * with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors may be used to
+ * endorse or promote products derived from this software without specific prior written
+ * permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.nodes.impl.intrinsics.interop;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.NodeChild;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.llvm.nodes.impl.base.LLVMAddressNode;
+import com.oracle.truffle.llvm.nodes.impl.base.LLVMLanguage;
+import com.oracle.truffle.llvm.nodes.impl.intrinsics.llvm.LLVMIntrinsic.LLVMAddressIntrinsic;
+import com.oracle.truffle.llvm.types.LLVMAddress;
+import com.oracle.truffle.llvm.types.LLVMTruffleObject;
+
+@NodeChild(type = LLVMAddressNode.class)
+public abstract class LLVMTruffleImportCached extends LLVMAddressIntrinsic {
+
+    @Child private CacheEntry cache = new UnresolvedCacheEntry();
+
+    @Specialization
+    public Object executeIntrinsic(LLVMAddress value) {
+        String id = LLVMTruffleIntrinsicUtil.readString(value);
+        return cache.execute(id);
+    }
+
+    // infinite cache implementation:
+
+    private static abstract class CacheEntry extends Node {
+        public abstract Object execute(String name);
+    }
+
+    private static final class UnresolvedCacheEntry extends CacheEntry {
+
+        @Override
+        public Object execute(String name) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            TruffleObject foreignObject = (TruffleObject) LLVMLanguage.INSTANCE.getEnvironment().importSymbol(name);
+            LLVMTruffleObject value = new LLVMTruffleObject(foreignObject);
+
+            Node n = this;
+            while (!(n instanceof LLVMTruffleImportCached)) {
+                n = n.getParent();
+            }
+            LLVMTruffleImportCached head = (LLVMTruffleImportCached) n;
+            head.cache = insert(new ResolvedCacheEntry(name, value, head.cache));
+            head.adoptChildren();
```

Comment:
I believe this should be `head.insert(...)` - probably the reason why you added `adoptChildren`?